### PR TITLE
First stab at reorganizing auth doc for GA

### PR DESCRIPTION
**Pull request by LisaFC**
As per https://github.com/grpc/grpc.github.io/issues/162

### CODE REVIEW

**Comment by LisaFC** [comment]
@jboeuf, feel free to link in anyone else who might be interested.  As per my mail to you we still need:
- some C++ examples (I know there are some in the general API section but they should be comprehensive and in Examples as well, as that's where users are going to look for them)
- some Objective C examples (there are none)
- just using SSL/TLS example for PHP (it looks like it's in the auth with Google example but I don't know enough about PHP to pull it out separately)
- Scopeless credentials example for Java (if it exists)

We should possibly also explain to users what the difference is between scoped and scopeless credentials for accessing Google services?

File: docs/guides/auth.md [hunk 1]
```diff
@@ -149,50 +152,159 @@ These authentication mechanisms will be available in all gRPC's supported langua
...
 stub = Helloworld::Greeter::Stub.new('localhost:50051', :this_channel_is_insecure)
 ...
+```
 
-# With server authentication SSL/TLS
+#### With server authentication SSL/TLS
+
+```ruby
 creds = GRPC::Core::Credentials.new(load_certs)  # load_certs typically loads a CA roots file
 stub = Helloworld::Greeter::Stub.new('localhost:50051', creds)
```
Comment (jboeuf) [inline docs/guides/auth.md]: 'myservice.example.com' as opposed to 'localhost:50051'

File: docs/guides/auth.md [hunk 2]
```diff
@@ -149,50 +152,159 @@ These authentication mechanisms will be available in all gRPC's supported langua
...
+require 'googleauth'  # from http://www.rubydoc.info/gems/googleauth/0.1.0
+...
+ssl_creds = GRPC::Core::ChannelCredentials.new(load_certs)  # load_certs typically loads a CA roots file
+authentication = Google::Auth.get_application_default()
+call_creds = GRPC::Core::CallCredentials.new(authentication.updater_proc)
+combined_creds = ssl_creds.compose(call_creds)
+stub = Helloworld::Greeter::Stub.new('greeter.googleapis.com', combined_creds)
+```
+
+#### Authenticate with Google using Oauth2 token (legacy approach)
```
Comment (jboeuf) [inline docs/guides/auth.md]: I think that we should remove this. @jtattermusch objections?

File: docs/guides/auth.md [hunk 3]
```diff
@@ -149,50 +152,159 @@ These authentication mechanisms will be available in all gRPC's supported langua
...
+...
+// Loads Google Application Default Credentials with publicly trusted roots.
+var channelCredentials = await GoogleGrpcCredentials.GetApplicationDefaultAsync();
+
+var channel = new Channel("greeter.googleapis.com", channelCredentials);
+var client = new Greeter.GreeterClient(channel);
+...
+```
+
+####Authenticate with Google using OAuth2 token (legacy approach)
```
Comment (jboeuf) [inline docs/guides/auth.md]: Remove this? @jtattermusch?

File: docs/guides/auth.md [hunk 4]
```diff
@@ -149,50 +152,159 @@ These authentication mechanisms will be available in all gRPC's supported langua
...
+if (googleCredential.IsCreateScopedRequired)
+{
+    googleCredential = googleCredential.CreateScoped(new[] { scope });
+}
+var channel = new Channel("greeter.googleapis.com", googleCredential.ToChannelCredentials());
+var client = new Greeter.GreeterClient(channel);
+...
+```
+
+####Authenticate a single RPC call
```
Comment (jboeuf) [inline docs/guides/auth.md]: We should probably add a section like this for the other languages as well.

File: docs/guides/auth.md [hunk 5]
```diff
@@ -149,50 +152,159 @@ These authentication mechanisms will be available in all gRPC's supported langua
...
+
+#### With server authentication SSL/TLS
+
+```python
+from grpc.beta import implementations
+import helloworld_pb2
 
-# With server authentication SSL/TLS
 creds = implementations.ssl_channel_credentials(open('roots.pem').read(), None, None)
 channel = implementations.secure_channel('localhost', 50051, creds)
```
Comment (jboeuf) [inline docs/guides/auth.md]: `'localhost', 50051` -> `'myservice.example.com', 443`

File: docs/guides/auth.md [hunk 6]
```diff
@@ -149,50 +152,159 @@ These authentication mechanisms will be available in all gRPC's supported langua
...
+    credentials = oauth2client.client.GoogleCredentials.get_application_default()
+    scoped_credentials = credentials.create_scoped([scope])
+  except Exception as error:
+    callback([], error)
+    return
+  callback([('authorization', 'Bearer %s' % scoped_credentials.get_access_token().access_token)], None)
+
+auth_creds = implementations.metadata_plugin_credentials(oauth2token_credentials)
+channel_creds = implementations.composite_channel_credentials(transport_creds, auth_creds)
+channel = implementations.secure_channel('localhost', 50051, channel_creds)
```
Comment (jboeuf) [inline docs/guides/auth.md]: `'localhost', 50051` -> `'greeter.googleapis.com', 443`

File: docs/guides/auth.md [hunk 7]
```diff
@@ -211,12 +323,6 @@ If the issuing certificate authority is not known to the client then a properly
...
 
 ```java
-// Base case - No encryption
-ManagedChannel channel = ManagedChannelBuilder.forAddress("localhost", 50051)
-    .usePlaintext(true)
-    .build();
-GreeterGrpc.GreeterStub stub = GreeterGrpc.newStub(channel);
-
 // With server authentication SSL/TLS
 ManagedChannel channel = ManagedChannelBuilder.forAddress("localhost", 50051)
```
Comment (jboeuf) [inline docs/guides/auth.md]: `'localhost', 50051` -> `'greeter.googleapis.com`

File: docs/guides/auth.md [hunk 8]
```diff
@@ -229,48 +335,42 @@ ManagedChannel channel = NettyChannelBuilder.forAddress("localhost", 50051)
...
 
-###Authenticating with Google (Node.js)
-
-####Base case - No encryption/authentication
+#### With server authentication SSL/TLS
 
 ```js
-var stub = new helloworld.Greeter('localhost:50051', grpc.credentials.createInsecure());
+var ssl_creds = grpc.credentials.createSsl(root_certs);
+var stub = new helloworld.Greeter('localhost:50051', ssl_creds);
```
Comment (jboeuf) [inline docs/guides/auth.md]: `'localhost:50051` -> `myservice.example.com`

**Comment by jboeuf** [comment]
Thanks much for the refactoring, it looks great. 
I added a few comments regarding target names so that we are consistent between languages and that these make sense security-wise:
- localhost for plaintext.
- myservice.example.com[:443] for SSL only.
- greeter.googleapis.com[:443] for google auth.

I also think that we should remove the non-scopeless credentials. @anthm, @jtattermusch please chime in if you disagree.
When this CL is done, we should ask the language owners to provide missing samples. I volunteer for the C++.

File: docs/guides/auth.md (hunk 1 above)
Comment (LisaFC) [inline docs/guides/auth.md]: Done.

File: docs/guides/auth.md (hunk 5 above)
Comment (LisaFC) [inline docs/guides/auth.md]: Done.

File: docs/guides/auth.md (hunk 6 above)
Comment (LisaFC) [inline docs/guides/auth.md]: Done.

File: docs/guides/auth.md (hunk 8 above)
Comment (LisaFC) [inline docs/guides/auth.md]: Done.

**Comment by LisaFC** [comment]
BTW, if anyone wants to see the current rendered version of this update, it's here: http://lcarey0-l.dub.corp.google.com:4000/docs/guides/auth.html

File: docs/guides/auth.md (hunk 3 above)
Comment (jtattermusch) [inline docs/guides/auth.md]: Can we keep this for a little longer? We've recently had some issues with JWT access to some google APIs and also not all languagues support JWT fully yet.

Also, I'd prefer if we removed the OAuth2 sections for all languages in a separate PR (with the description of the PR stating why we are removing these).

**Comment by jboeuf** [comment]
Thanks for the fixes. LGTM for this part. Do you want me to merge or should we wait for the other items to be completed against this PR?

**Comment by LisaFC** [comment]
I'm happy to merge for now, I'll create new issues and/or PRs for the other bits we need to do to it.

**Pull request merged**